# Introducción a Transformers



## Instalar librerías

In [ ]:
!pip install transformers
!pip install datasets

## Importar Pipeline

Cargar la funcion pipeline que permite invocar las distintas operaciones de NLP

In [ ]:
from transformers import pipeline

## Modelos de NLP

Hay algunos pasos que son parte del proceso de interpretación dentro de la función pipeline y sus modelos:

1- El texto es pre-procesado en un formato que el modelo pueda entender\
2- Los textos pre-procesados entran al modelo\
3- La prediction del modelo es post-procesada, así tu puedes entender el resultado.

### Zero-shot classification

In [ ]:
classifier = pipeline("zero-shot-classification")
classifier(
    "This is a course about the Transformers library",
    candidate_labels=["education", "politics", "business"],
)

No model was supplied, defaulted to facebook/bart-large-mnli and revision d7645e1 (https://huggingface.co/facebook/bart-large-mnli).
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json: 0.00B [00:00, ?B/s]

model.safetensors:   0%|          | 0.00/1.63G [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

Device set to use cuda:0


{'sequence': 'This is a course about the Transformers library',
 'labels': ['education', 'business', 'politics'],
 'scores': [0.844595193862915, 0.11197695881128311, 0.04342786595225334]}

In [ ]:
classifier = pipeline("zero-shot-classification", model="tasksource/deberta-small-long-nli")
classifier(
    "This is a course about the Transformers library",
    candidate_labels=["education", "politics", "business"],
)

config.json: 0.00B [00:00, ?B/s]

model.safetensors:   0%|          | 0.00/568M [00:00<?, ?B/s]

tokenizer_config.json: 0.00B [00:00, ?B/s]

spm.model:   0%|          | 0.00/2.46M [00:00<?, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

added_tokens.json:   0%|          | 0.00/23.0 [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/286 [00:00<?, ?B/s]

Device set to use cuda:0
Asking to truncate to max_length but no maximum length is provided and the model has no predefined maximum length. Default to no truncation.


{'sequence': 'This is a course about the Transformers library',
 'labels': ['education', 'business', 'politics'],
 'scores': [0.81794673204422, 0.16893085837364197, 0.013122416101396084]}

Este pipeline es llamada zero-shot porque no necesitas hacer ningun fine-tunning del modelo sobre tus datos para usarlo. El puede calificar directamente la lista de etiquetas que tu desees evaluar.


In [ ]:
from transformers import pipeline
classifier = pipeline("zero-shot-classification", model="Xenova/nli-deberta-v3-xsmall")

sequence_to_classify = "Tribunal de Caldas negó demanda contra Presidencia por transmisión de consejo de ministros"
candidate_labels = ["politica", "economia", "entretenimiento", "ambiente"]
output = classifier(sequence_to_classify, candidate_labels, multi_label=False)
print(output)

config.json: 0.00B [00:00, ?B/s]

ValueError: Could not load model Xenova/nli-deberta-v3-xsmall with any of the following classes: (<class 'transformers.models.auto.modeling_auto.AutoModelForSequenceClassification'>, <class 'transformers.models.auto.modeling_tf_auto.TFAutoModelForSequenceClassification'>, <class 'transformers.models.deberta_v2.modeling_deberta_v2.DebertaV2ForSequenceClassification'>, <class 'transformers.models.deberta_v2.modeling_tf_deberta_v2.TFDebertaV2ForSequenceClassification'>). See the original errors:

while loading with AutoModelForSequenceClassification, an error is thrown:
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/transformers/pipelines/base.py", line 293, in infer_framework_load_model
    model = model_class.from_pretrained(model, **kwargs)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/transformers/models/auto/auto_factory.py", line 604, in from_pretrained
    return model_class.from_pretrained(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/transformers/modeling_utils.py", line 288, in _wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/transformers/modeling_utils.py", line 5027, in from_pretrained
    checkpoint_files, sharded_metadata = _get_resolved_checkpoint_files(
                                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/transformers/modeling_utils.py", line 1256, in _get_resolved_checkpoint_files
    raise OSError(
OSError: Xenova/nli-deberta-v3-xsmall does not appear to have a file named pytorch_model.bin, model.safetensors, tf_model.h5, model.ckpt or flax_model.msgpack.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/transformers/pipelines/base.py", line 311, in infer_framework_load_model
    model = model_class.from_pretrained(model, **fp32_kwargs)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/transformers/models/auto/auto_factory.py", line 604, in from_pretrained
    return model_class.from_pretrained(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/transformers/modeling_utils.py", line 288, in _wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/transformers/modeling_utils.py", line 5027, in from_pretrained
    checkpoint_files, sharded_metadata = _get_resolved_checkpoint_files(
                                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/transformers/modeling_utils.py", line 1256, in _get_resolved_checkpoint_files
    raise OSError(
OSError: Xenova/nli-deberta-v3-xsmall does not appear to have a file named pytorch_model.bin, model.safetensors, tf_model.h5, model.ckpt or flax_model.msgpack.

while loading with TFAutoModelForSequenceClassification, an error is thrown:
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/transformers/pipelines/base.py", line 293, in infer_framework_load_model
    model = model_class.from_pretrained(model, **kwargs)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/transformers/models/auto/auto_factory.py", line 604, in from_pretrained
    return model_class.from_pretrained(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/transformers/modeling_tf_utils.py", line 2857, in from_pretrained
    raise OSError(
OSError: Xenova/nli-deberta-v3-xsmall does not appear to have a file named pytorch_model.bin, tf_model.h5 or model.ckpt

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/transformers/pipelines/base.py", line 311, in infer_framework_load_model
    model = model_class.from_pretrained(model, **fp32_kwargs)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/transformers/models/auto/auto_factory.py", line 604, in from_pretrained
    return model_class.from_pretrained(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/transformers/modeling_tf_utils.py", line 2857, in from_pretrained
    raise OSError(
OSError: Xenova/nli-deberta-v3-xsmall does not appear to have a file named pytorch_model.bin, tf_model.h5 or model.ckpt

while loading with DebertaV2ForSequenceClassification, an error is thrown:
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/transformers/pipelines/base.py", line 293, in infer_framework_load_model
    model = model_class.from_pretrained(model, **kwargs)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/transformers/modeling_utils.py", line 288, in _wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/transformers/modeling_utils.py", line 5027, in from_pretrained
    checkpoint_files, sharded_metadata = _get_resolved_checkpoint_files(
                                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/transformers/modeling_utils.py", line 1256, in _get_resolved_checkpoint_files
    raise OSError(
OSError: Xenova/nli-deberta-v3-xsmall does not appear to have a file named pytorch_model.bin, model.safetensors, tf_model.h5, model.ckpt or flax_model.msgpack.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/transformers/pipelines/base.py", line 311, in infer_framework_load_model
    model = model_class.from_pretrained(model, **fp32_kwargs)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/transformers/modeling_utils.py", line 288, in _wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/transformers/modeling_utils.py", line 5027, in from_pretrained
    checkpoint_files, sharded_metadata = _get_resolved_checkpoint_files(
                                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/transformers/modeling_utils.py", line 1256, in _get_resolved_checkpoint_files
    raise OSError(
OSError: Xenova/nli-deberta-v3-xsmall does not appear to have a file named pytorch_model.bin, model.safetensors, tf_model.h5, model.ckpt or flax_model.msgpack.

while loading with TFDebertaV2ForSequenceClassification, an error is thrown:
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/transformers/pipelines/base.py", line 293, in infer_framework_load_model
    model = model_class.from_pretrained(model, **kwargs)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/transformers/modeling_tf_utils.py", line 2857, in from_pretrained
    raise OSError(
OSError: Xenova/nli-deberta-v3-xsmall does not appear to have a file named pytorch_model.bin, tf_model.h5 or model.ckpt

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/transformers/pipelines/base.py", line 311, in infer_framework_load_model
    model = model_class.from_pretrained(model, **fp32_kwargs)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/transformers/modeling_tf_utils.py", line 2857, in from_pretrained
    raise OSError(
OSError: Xenova/nli-deberta-v3-xsmall does not appear to have a file named pytorch_model.bin, tf_model.h5 or model.ckpt




In [ ]:
from transformers import pipeline
classifier = pipeline("zero-shot-classification", model="MoritzLaurer/mDeBERTa-v3-base-mnli-xnli")

sequence_to_classify = "Tribunal de Caldas negó demanda contra Presidencia por transmisión de consejo de ministros"
candidate_labels = ["politica", "economia", "entretenimiento", "ambiente"]
output = classifier(sequence_to_classify, candidate_labels, multi_label=False)
print(output)


config.json: 0.00B [00:00, ?B/s]

model.safetensors:   0%|          | 0.00/558M [00:00<?, ?B/s]

tokenizer_config.json: 0.00B [00:00, ?B/s]

spm.model:   0%|          | 0.00/4.31M [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/16.3M [00:00<?, ?B/s]

added_tokens.json:   0%|          | 0.00/23.0 [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/286 [00:00<?, ?B/s]

Device set to use cuda:0


{'sequence': 'Tribunal de Caldas negó demanda contra Presidencia por transmisión de consejo de ministros', 'labels': ['politica', 'economia', 'entretenimiento', 'ambiente'], 'scores': [0.83127760887146, 0.06993889808654785, 0.0625704973936081, 0.03621302545070648]}



### Clasificación de textos

In [ ]:
Classification = pipeline('text-classification', model='nlptown/bert-base-multilingual-uncased-sentiment')

config.json:   0%|          | 0.00/953 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/669M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/39.0 [00:00<?, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

Device set to use cuda:0


In [ ]:
Classification('this class is horrible')

[{'label': '1 star', 'score': 0.9311068654060364}]

In [ ]:
Classification('this class is amazing')

[{'label': '5 stars', 'score': 0.8783395290374756}]

In [ ]:
classi = pipeline('text-classification', model='nlptown/bert-base-multilingual-uncased-sentiment')

In [ ]:
classi('la clase es buena')

[{'label': '4 stars', 'score': 0.46930399537086487}]

### Fill-mask

In [ ]:
unmasker = pipeline("fill-mask")
unmasker("This course will teach you all about <mask> models.", top_k=5)

No model was supplied, defaulted to distilbert/distilroberta-base and revision fb53ab8 (https://huggingface.co/distilbert/distilroberta-base).
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json:   0%|          | 0.00/480 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/331M [00:00<?, ?B/s]

Some weights of the model checkpoint at distilbert/distilroberta-base were not used when initializing RobertaForMaskedLM: ['roberta.pooler.dense.bias', 'roberta.pooler.dense.weight']
- This IS expected if you are initializing RobertaForMaskedLM from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing RobertaForMaskedLM from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).


tokenizer_config.json:   0%|          | 0.00/25.0 [00:00<?, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

Device set to use cuda:0


[{'score': 0.19619838893413544,
  'token': 30412,
  'token_str': ' mathematical',
  'sequence': 'This course will teach you all about mathematical models.'},
 {'score': 0.04052715003490448,
  'token': 38163,
  'token_str': ' computational',
  'sequence': 'This course will teach you all about computational models.'},
 {'score': 0.03301798552274704,
  'token': 27930,
  'token_str': ' predictive',
  'sequence': 'This course will teach you all about predictive models.'},
 {'score': 0.031941574066877365,
  'token': 745,
  'token_str': ' building',
  'sequence': 'This course will teach you all about building models.'},
 {'score': 0.024522822350263596,
  'token': 3034,
  'token_str': ' computer',
  'sequence': 'This course will teach you all about computer models.'}]

In [ ]:
mask = pipeline('fill-mask', model='bert-base-uncased')

config.json:   0%|          | 0.00/570 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/440M [00:00<?, ?B/s]

Some weights of the model checkpoint at bert-base-uncased were not used when initializing BertForMaskedLM: ['bert.pooler.dense.bias', 'bert.pooler.dense.weight', 'cls.seq_relationship.bias', 'cls.seq_relationship.weight']
- This IS expected if you are initializing BertForMaskedLM from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertForMaskedLM from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).


tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

In [ ]:
pred = mask('This course will teach you all about [MASK] models.')

In [ ]:
for predict in pred:
  print(f"Predicted token: {predict['token_str']} | Confidence: {predict['score']: .4f}")
#

Predicted token: role | Confidence:  0.3366
Predicted token: business | Confidence:  0.0837
Predicted token: the | Confidence:  0.0572
Predicted token: your | Confidence:  0.0439
Predicted token: fashion | Confidence:  0.0267


### Text-generation

In [ ]:
generator = pipeline('text-generation', model='gpt2')

config.json:   0%|          | 0.00/665 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/548M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/124 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/1.04M [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]

Device set to use cuda:0


In [ ]:
import torch
from transformers import pipeline

model_id = "meta-llama/Llama-3.2-1B"

pipe = pipeline(
    "text-generation",
    model=model_id,
    torch_dtype=torch.bfloat16,
    device_map="auto"
)

pipe("The key to life is")


OSError: You are trying to access a gated repo.
Make sure to have access to it at https://huggingface.co/meta-llama/Llama-3.2-1B.
403 Client Error. (Request ID: Root=1-67e7436d-29c976770b5febdc7d292e56;4372c739-1c32-4c26-931b-6253dc537980)

Cannot access gated repo for url https://huggingface.co/meta-llama/Llama-3.2-1B/resolve/main/config.json.
Access to model meta-llama/Llama-3.2-1B is restricted and you are not in the authorized list. Visit https://huggingface.co/meta-llama/Llama-3.2-1B to ask for access.

In [ ]:
generator('The key to life is')

Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


[{'generated_text': 'The key to life is for any man to be a man.\n\n"And as for me, I\'m not a man to be a woman," said the young woman, who is in her 20s. While the man was there to celebrate her victory over the other contestants, she said she wasn\'t angry.\n\n"I\'m a man with a family," she said.\n\nA woman who asked that her last name not be used.\n\n"I\'m a woman," she said. "I\'m a woman. I\'m a woman. I\'m a woman."\n\nA woman who is a member of the U.S. military.\n\nA woman who was in the military when she was shot and remains hospitalized with injuries, who said she was treated in hospital on the morning of Monday.\n\n"I\'m not a woman," she said. "I\'m a woman. I\'m a woman. I\'m a woman."\n\nIn a statement, the U.S. military said it is monitoring the situation and will provide updates as soon as possible.\n\nThe incident was reported at 8:30 a.m. CST, according to the U.S. military.\n\n"We are cooperating with law enforcement in the investigation and will'}]

In [ ]:
generator('El dia esta soleado')

Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


[{'generated_text': 'El dia esta soleado:\n\nO se llegado.\n\nO se llegado:\n\nO llegado:\n\nO se llegado:\n\nO se llegado:\n\nO se llegado:\n\nO se llegado:\n\nO se llegado:\n\nO se llegado:\n\nO se llegado:\n\nO se llegado:\n\nO se llegado:\n\nO se llegado:\n\nO se llegado:\n\nO se llegado:\n\nO se llegado:\n\nO se llegado:\n\nO se llegado:\n\nO se llegado:\n\nO se llegado:\n\nO se llegado:\n\nO se llegado:\n\nO se llegado:\n\nO se llegado:\n\nO se llegado:\n\nO se llegado:\n\nO se llegado:\n\nO se llegado:\n\nO se llegado:\n\nO se llegado:\n\nO se llegado:\n\nO se llegado:'}]

In [ ]:
generator = pipeline("text-generation")
generator("In this course, we will teach you how to")

No model was supplied, defaulted to openai-community/gpt2 and revision 6c0e608 (https://huggingface.co/openai-community/gpt2).
Using a pipeline without specifying a model name and revision in production is not recommended.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


[{'generated_text': 'In this course, we will teach you how to use your computer with the touch screen to create a virtual globe. However, once you understand how to use the touchscreen to make a virtual globe, you will be able to use the touch screen to generate'}]

In [ ]:
text = '''
Hey Jack, I just wanted to flag something with you. Last week when you
said that you didn't want to watch the move Twilight with me, even in jest, it kind
of got under my skin. I mainly feel like it's hypocritical when you make me watch basketball
games with you and our main activity together is watching sports on TV. I just wanted to get
it off my chest. From Sophie
'''

In [ ]:
generator = pipeline(
    "text-generation",
)
prompt = text + "\n\nJack apology to Sophie:\n\nDear Sophie, I am sorry I didn't"
outputs = generator(prompt, max_length=200)
print(outputs[0]['generated_text'])

No model was supplied, defaulted to openai-community/gpt2 and revision 607a30d (https://huggingface.co/openai-community/gpt2).
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json:   0%|          | 0.00/665 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/548M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/124 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/1.04M [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]

Device set to use cuda:0
Truncation was not explicitly activated but `max_length` is provided a specific value, please use `truncation=True` to explicitly truncate examples to max length. Defaulting to 'longest_first' truncation strategy. If you encode pairs of sequences (GLUE-style) with the tokenizer you can select this strategy more precisely by providing a specific strategy to `truncation`.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.



Hey Jack, I just wanted to flag something with you. Last week when you
said that you didn't want to watch the move Twilight with me, even in jest, it kind
of got under my skin. I mainly feel like it's hypocritical when you make me watch basketball
games with you and our main activity together is watching sports on TV. I just wanted to get
it off my chest. From Sophie


Jack apology to Sophie:

Dear Sophie, I am sorry I didn't clarify enough.

I only told about this because I didn't want people thinking that

I had to worry about myself. It wouldn't make any sense to me.

I never intended to cause trouble for any of our fans either, so I am sorry to see

you lose your relationship with so many people.

Sophie, you are the reason that all of this occurred. You are my new best friend,

just in


Puedes controlar cuantas secuencias diferentes son generadas con el argumento *num_return_sequences* y la longitud total de la salida de texto con el argumento *max_length*

**Intentalo**: Use los argumentos indicados para generar dos oraciones de 15 palabras cada uno.

In [ ]:
generator = pipeline("text-generation", model="distilgpt2")
generator(
    "In this course you learn about",
    max_length=15,
    num_return_sequences=2,
)

Truncation was not explicitly activated but `max_length` is provided a specific value, please use `truncation=True` to explicitly truncate examples to max length. Defaulting to 'longest_first' truncation strategy. If you encode pairs of sequences (GLUE-style) with the tokenizer you can select this strategy more precisely by providing a specific strategy to `truncation`.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


[{'generated_text': 'In this course you learn about many fundamental elements such as the right to bear'},
 {'generated_text': 'In this course you learn about how to use a Java program to build a'}]

In [ ]:
generator = pipeline("text-generation", model="distilgpt2")
generator(
    "In this course, we will teach you how to",
    max_length=30,
    num_return_sequences=2,
)

Truncation was not explicitly activated but `max_length` is provided a specific value, please use `truncation=True` to explicitly truncate examples to max length. Defaulting to 'longest_first' truncation strategy. If you encode pairs of sequences (GLUE-style) with the tokenizer you can select this strategy more precisely by providing a specific strategy to `truncation`.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


[{'generated_text': 'In this course, we will teach you how to achieve success as we have always been proud of our knowledge, especially the skills people have given them.'},
 {'generated_text': 'In this course, we will teach you how to create a new game called The Adventure in 3D. How do you complete the process, by completing'}]

**Intentalo**:  Use los filtros dentro de la página de HuggingFace para encontrar un modelo de generación de texto para otro lenguaje. Prueba el modelo seleccionado invocandolo dentro del operador pipeline.

In [ ]:
# Prueba con generador de texto en español "gagan3012/Multilingual-mistral"
generator = pipeline("text-generation", model="gagan3012/Multilingual-mistral")
generator(
    "El dia de mañana me voy a"
)

config.json:   0%|          | 0.00/773 [00:00<?, ?B/s]

model.safetensors.index.json:   0%|          | 0.00/88.7k [00:00<?, ?B/s]

model-00001-of-00010.safetensors:   0%|          | 0.00/9.92G [00:00<?, ?B/s]

KeyboardInterrupt: 

### Sentiment analysis

In [ ]:
classifier = pipeline("sentiment-analysis")
classifier("La película tiene una trama muy interesante pero el desarrollo de los personajes fue muy lento.")

No model was supplied, defaulted to distilbert/distilbert-base-uncased-finetuned-sst-2-english and revision 714eb0f (https://huggingface.co/distilbert/distilbert-base-uncased-finetuned-sst-2-english).
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json:   0%|          | 0.00/629 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/268M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

Device set to use cuda:0


[{'label': 'NEGATIVE', 'score': 0.97259122133255}]

In [ ]:
classifier(
    ["I've been waiting for a Machine Learning course my whole life.", "I hate this so much!"]
)

[{'label': 'NEGATIVE', 'score': 0.9948803186416626},
 {'label': 'NEGATIVE', 'score': 0.9994558691978455}]

In [ ]:
classifier = pipeline(
    "text-classification",
    model="distilbert-base-uncased-finetuned-sst-2-english",
)
outputs = classifier(text)
outputs

[{'label': 'NEGATIVE', 'score': 0.912985622882843}]

### Named Entity Recognition NER


In [ ]:
ner = pipeline("ner", grouped_entities=True)
ner("My name is Sylvain and I work at The Herald in Brooklyn.")

No model was supplied, defaulted to dbmdz/bert-large-cased-finetuned-conll03-english and revision 4c53496 (https://huggingface.co/dbmdz/bert-large-cased-finetuned-conll03-english).
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json:   0%|          | 0.00/998 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/1.33G [00:00<?, ?B/s]

Some weights of the model checkpoint at dbmdz/bert-large-cased-finetuned-conll03-english were not used when initializing BertForTokenClassification: ['bert.pooler.dense.bias', 'bert.pooler.dense.weight']
- This IS expected if you are initializing BertForTokenClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertForTokenClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).


tokenizer_config.json:   0%|          | 0.00/60.0 [00:00<?, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

Device set to use cuda:0
/usr/local/lib/python3.12/dist-packages/transformers/pipelines/token_classification.py:186: UserWarning: `grouped_entities` is deprecated and will be removed in version v5.0.0, defaulted to `aggregation_strategy="AggregationStrategy.SIMPLE"` instead.
  warnings.warn(


[{'entity_group': 'PER',
  'score': np.float32(0.9986294),
  'word': 'Sylvain',
  'start': 11,
  'end': 18},
 {'entity_group': 'ORG',
  'score': np.float32(0.9975711),
  'word': 'The Herald',
  'start': 33,
  'end': 43},
 {'entity_group': 'LOC',
  'score': np.float32(0.9848874),
  'word': 'Brooklyn',
  'start': 47,
  'end': 55}]

In [ ]:
ner_tagger = pipeline(
    "ner",
    aggregation_strategy="simple",
    model="dbmdz/bert-large-cased-finetuned-conll03-english",
)
outputs = ner_tagger(text)
outputs

Some weights of the model checkpoint at dbmdz/bert-large-cased-finetuned-conll03-english were not used when initializing BertForTokenClassification: ['bert.pooler.dense.bias', 'bert.pooler.dense.weight']
- This IS expected if you are initializing BertForTokenClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertForTokenClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Device set to use cuda:0


[{'entity_group': 'PER',
  'score': np.float32(0.9891127),
  'word': 'Jack',
  'start': 5,
  'end': 9},
 {'entity_group': 'MISC',
  'score': np.float32(0.99652916),
  'word': 'Twilight',
  'start': 116,
  'end': 124},
 {'entity_group': 'PER',
  'score': np.float32(0.61472476),
  'word': 'Sophie',
  'start': 364,
  'end': 370}]

### Question answering

El pipeline de *question-answering* responde preguntas usando la información dad con algún contexto:

In [ ]:
question_answerer = pipeline("question-answering")
question_answerer(
    question="Where do I work?",
    context="My name is Sylvain and I work at Hugging Face in Brooklyn",
)

No model was supplied, defaulted to distilbert/distilbert-base-cased-distilled-squad and revision 564e9b5 (https://huggingface.co/distilbert/distilbert-base-cased-distilled-squad).
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json:   0%|          | 0.00/473 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/261M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/49.0 [00:00<?, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

Fetching 0 files: 0it [00:00, ?it/s]

Fetching 1 files:   0%|          | 0/1 [00:00<?, ?it/s]

Fetching 0 files: 0it [00:00, ?it/s]

Device set to use cuda:0


{'score': 0.6949763894081116, 'start': 33, 'end': 45, 'answer': 'Hugging Face'}

In [ ]:
reader = pipeline(
    "question-answering",
    model="distilbert-base-cased-distilled-squad",
)
question = "What movie did Jack not watch?"
outputs = reader(question=question, context=text)
outputs

config.json:   0%|          | 0.00/473 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/261M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/49.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/213k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/436k [00:00<?, ?B/s]

Fetching 0 files: 0it [00:00, ?it/s]

Fetching 1 files:   0%|          | 0/1 [00:00<?, ?it/s]

Fetching 0 files: 0it [00:00, ?it/s]

Device set to use cuda:0


{'score': 0.9831923246383667, 'start': 116, 'end': 124, 'answer': 'Twilight'}

### Summarization

La sumarizacion es la tarea que busca resumir un texto conservando los aspectos más relevantes del texto original. Ejemplo:


In [ ]:
summarizer = pipeline("summarization")
summarizer(
    """
    America has changed dramatically during recent years. Not only has the number of
    graduates in traditional engineering disciplines such as mechanical, civil,
    electrical, chemical, and aeronautical engineering declined, but in most of
    the premier American universities engineering curricula now concentrate on
    and encourage largely the study of engineering science. As a result, there
    are declining offerings in engineering subjects dealing with infrastructure,
    the environment, and related issues, and greater concentration on high
    technology subjects, largely supporting increasingly complex scientific
    developments. While the latter is important, it should not be at the expense
    of more traditional engineering.

    Rapidly developing economies such as China and India, as well as other
    industrial countries in Europe and Asia, continue to encourage and advance
    the teaching of engineering. Both China and India, respectively, graduate
    six and eight times as many traditional engineers as does the United States.
    Other industrial countries at minimum maintain their output, while America
    suffers an increasingly serious decline in the number of engineering graduates
    and a lack of well-educated engineers.
"""
)

No model was supplied, defaulted to sshleifer/distilbart-cnn-12-6 and revision a4f8f3e (https://huggingface.co/sshleifer/distilbart-cnn-12-6).
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json:   0%|          | 0.00/1.80k [00:00<?, ?B/s]

pytorch_model.bin:   0%|          | 0.00/1.22G [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/1.22G [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/899k [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

Device set to use cuda:0


[{'summary_text': ' The number of engineering graduates in the United States has declined in recent years . China and India graduate six and eight times as many traditional engineers as the U.S. does . Rapidly developing economies such as China continue to encourage and advance the teaching of engineering . There are declining offerings in engineering subjects dealing with infrastructure, infrastructure, the environment, and related issues .'}]

In [ ]:
summarizer = pipeline(
    "summarization",
    model="sshleifer/distilbart-cnn-12-6",
)
outputs = summarizer(text, max_length=60, clean_up_tokenization_spaces=True)
print(outputs[0]['summary_text'])

 Sophie: "I just wanted to get  it off my chest. I mainly feel like it's hypocritical when you make me watch basketball games with you and our main activity together is watching sports on TV" Sophie says she feels hypocritical when she and Jack watch sports together.


### Translation

Para tareas de *traducción* se puede invocar el modelo base si se especifica el par de lenguajes que se quieren usar (ejemplo: "translation_en_to_fr"),
pero es más facilo seleccionar el modelo desde el Model Hub.

In [ ]:
translator = pipeline("translation", model="Helsinki-NLP/opus-mt-fr-es")
translator("Ce cours est produit par Facebook.")

config.json: 0.00B [00:00, ?B/s]

pytorch_model.bin:   0%|          | 0.00/332M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/293 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/332M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/42.0 [00:00<?, ?B/s]

source.spm:   0%|          | 0.00/812k [00:00<?, ?B/s]

target.spm:   0%|          | 0.00/819k [00:00<?, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

/usr/local/lib/python3.12/dist-packages/transformers/models/marian/tokenization_marian.py:175: UserWarning: Recommended: pip install sacremoses.
  warnings.warn("Recommended: pip install sacremoses.")
Device set to use cuda:0


[{'translation_text': 'Este curso es producido por Facebook.'}]

**Intentalo**:  Busca por **translation models** en otros lenguajes y trata de traducir la frase anterior en 5 idiomas distintos.

## Usando Transformers

Veamos lo que sucede detrás de escena al usar aguno de los modelos de pipeline anteriormente descritos

In [ ]:
from transformers import pipeline

classifier = pipeline("sentiment-analysis")
classifier(
    [
        "I've been waiting for a Machine Learning course my whole life.",
        "I hate this so much!",
    ]
)

No model was supplied, defaulted to distilbert/distilbert-base-uncased-finetuned-sst-2-english and revision af0f99b (https://huggingface.co/distilbert/distilbert-base-uncased-finetuned-sst-2-english).
Using a pipeline without specifying a model name and revision in production is not recommended.


[{'label': 'NEGATIVE', 'score': 0.9948803186416626},
 {'label': 'NEGATIVE', 'score': 0.9994558691978455}]

### Preprocesamiento con un tokenizador
Al igual que otras redes neuronales, los modelos **Transformer** no pueden procesar texto sin procesar directamente, por lo que el primer paso de nuestra canalización es convertir las entradas de texto en números que el modelo pueda entender. Para ello utilizamos un tokenizador, que se encargará de:

1.   Dividir la entrada en palabras, subpalabras o símbolos (como signos de puntuación) que se denominan tokens.
2.   Asignación de cada token a un número entero.
3.   Adición de entradas adicionales que pueden ser útiles para el modelo.

Todo este preprocesamiento debe realizarse exactamente de la misma manera que cuando se entrenó previamente el modelo, por lo que primero debemos descargar esa información del Model Hub. Para ello, utilizamos la clase AutoTokenizer y su método from_pretrained(). Usando el nombre del punto de control de nuestro modelo, recuperará automáticamente los datos asociados con el tokenizador del modelo y los almacenará en caché (por lo que solo se descarga la primera vez que ejecuta el código a continuación).

In [ ]:
from transformers import AutoTokenizer

checkpoint = "distilbert-base-uncased-finetuned-sst-2-english"
tokenizer = AutoTokenizer.from_pretrained(checkpoint)

In [ ]:
tokenizer

DistilBertTokenizerFast(name_or_path='distilbert-base-uncased-finetuned-sst-2-english', vocab_size=30522, model_max_length=512, is_fast=True, padding_side='right', truncation_side='right', special_tokens={'unk_token': '[UNK]', 'sep_token': '[SEP]', 'pad_token': '[PAD]', 'cls_token': '[CLS]', 'mask_token': '[MASK]'}, clean_up_tokenization_spaces=True),  added_tokens_decoder={
	0: AddedToken("[PAD]", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
	100: AddedToken("[UNK]", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
	101: AddedToken("[CLS]", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
	102: AddedToken("[SEP]", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
	103: AddedToken("[MASK]", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
}

Una vez que tengamos el tokenizador, podemos pasarle directamente nuestras oraciones y obtendremos un diccionario que está listo para alimentar a nuestro modelo. Lo único que queda por hacer es convertir la lista de ID de entrada en tensores.

Puede usar 🤗 Transformers sin tener que preocuparse por qué marco de ML se usa como backend; puede ser PyTorch o TensorFlow, o Flax para algunos modelos. Sin embargo, los modelos Transformer solo aceptan tensores como entrada.

In [ ]:
raw_inputs = [
    "I've been waiting for a Machine Learning course my whole life.",
    "I hate this so much!",
]
inputs = tokenizer(raw_inputs, padding=True, truncation=True, return_tensors="pt")
print(inputs)

{'input_ids': tensor([[ 101, 1045, 1005, 2310, 2042, 3403, 2005, 1037, 3698, 4083, 2607, 2026,
         2878, 2166, 1012,  102],
        [ 101, 1045, 5223, 2023, 2061, 2172,  999,  102,    0,    0,    0,    0,
            0,    0,    0,    0]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
        [1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0]])}


La salida en sí es un diccionario que contiene dos claves, input_ids y attention_mask. input_ids contiene dos filas de enteros (uno para cada frase) que son los identificadores únicos de los tokens de cada frase.

Podemos descargar nuestro modelo preentrenado de la misma manera que lo hicimos con nuestro tokenizador. 🤗 Transformers proporciona una clase AutoModel que también tiene un método from_pretrained():

In [ ]:
from transformers import AutoModel

checkpoint = "distilbert-base-uncased-finetuned-sst-2-english"
model = AutoModel.from_pretrained(checkpoint)

In [ ]:
model

DistilBertModel(
  (embeddings): Embeddings(
    (word_embeddings): Embedding(30522, 768, padding_idx=0)
    (position_embeddings): Embedding(512, 768)
    (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
    (dropout): Dropout(p=0.1, inplace=False)
  )
  (transformer): Transformer(
    (layer): ModuleList(
      (0-5): 6 x TransformerBlock(
        (attention): MultiHeadSelfAttention(
          (dropout): Dropout(p=0.1, inplace=False)
          (q_lin): Linear(in_features=768, out_features=768, bias=True)
          (k_lin): Linear(in_features=768, out_features=768, bias=True)
          (v_lin): Linear(in_features=768, out_features=768, bias=True)
          (out_lin): Linear(in_features=768, out_features=768, bias=True)
        )
        (sa_layer_norm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
        (ffn): FFN(
          (dropout): Dropout(p=0.1, inplace=False)
          (lin1): Linear(in_features=768, out_features=3072, bias=True)
          (lin2): Li

Esta arquitectura contiene solo el módulo base Transformer: dadas algunas entradas, emite lo que llamaremos estados ocultos, también conocidos como características. Para cada entrada del modelo, recuperaremos un vector de alta dimensión que representa la comprensión contextual de esa entrada por parte del modelo Transformer.

La salida vectorial del módulo Transformer suele ser grande. Generalmente tiene tres dimensiones:

1. Tamaño de lote: El número de secuencias procesadas a la vez (2 en nuestro ejemplo).
2. Longitud de la secuencia: La longitud de la representación numérica de la secuencia (16 en nuestro ejemplo).
3. Tamaño oculto: la dimensión vectorial de cada entrada del modelo.

Se dice que es de "alta dimensión" debido al último valor. El tamaño oculto puede ser muy grande (768 es común para los modelos más pequeños, y en los modelos más grandes puede llegar a 3072 o más).

In [ ]:
outputs = model(**inputs)
print(outputs.last_hidden_state.shape)

torch.Size([2, 16, 768])


Hay muchas arquitecturas diferentes disponibles en 🤗 Transformers, cada una diseñada para abordar una tarea específica. He aquí una lista no exhaustiva:

*  Model (retrieve the hidden states)
*  ForCausalLM
*  ForMaskedLM
*  ForMultipleChoice
*  ForQuestionAnswering
*  ForSequenceClassification
*  ForTokenClassification



Para nuestro ejemplo, necesitaremos un modelo con un encabezado de clasificación de secuencia (para poder clasificar las oraciones como positivas o negativas). Por lo tanto, en realidad no usaremos la clase AutoModel, sino AutoModelForSequenceClassification:

In [ ]:
from transformers import AutoModelForSequenceClassification

checkpoint = "distilbert-base-uncased-finetuned-sst-2-english"
model = AutoModelForSequenceClassification.from_pretrained(checkpoint)
outputs = model(**inputs)

In [ ]:
outputs

SequenceClassifierOutput(loss=None, logits=tensor([[ 2.8205, -2.4491],
        [ 4.1692, -3.3464]], grad_fn=<AddmmBackward0>), hidden_states=None, attentions=None)

Ahora, si nos fijamos en la forma de nuestras salidas, la dimensionalidad será mucho menor: la cabeza del modelo toma como entrada los vectores de alta dimensión que vimos antes, y genera vectores que contienen dos valores (uno por etiqueta):

In [ ]:
print(outputs.logits.shape)

torch.Size([2, 2])


Dado que tenemos solo dos oraciones y dos etiquetas, el resultado que obtenemos de nuestro modelo es de forma 2 x 2.

**Proecesando la salida del modelo**
Los valores que obtenemos como resultado de nuestro modelo no necesariamente tienen sentido por sí mismos. Echemos un vistazo:

In [ ]:
print(outputs.logits)

tensor([[ 2.8205, -2.4491],
        [ 4.1692, -3.3464]], grad_fn=<AddmmBackward0>)


 No se trata de probabilidades, sino de logits, las puntuaciones brutas y no normalizadas generadas por la última capa del modelo. Para convertirse en probabilidades, deben pasar por una capa SoftMax (todos los 🤗 modelos de Transformers emiten los logits, ya que la función de pérdida para el entrenamiento generalmente fusionará la última función de activación, como SoftMax, con la función de pérdida real, como la entropía cruzada):

In [ ]:
import torch

predictions = torch.nn.functional.softmax(outputs.logits, dim=-1)
print(predictions)

tensor([[9.9488e-01, 5.1197e-03],
        [9.9946e-01, 5.4418e-04]], grad_fn=<SoftmaxBackward0>)


 Para obtener las etiquetas correspondientes a cada posición, podemos inspeccionar el atributo id2label de la configuración del modelo:

In [ ]:
model.config.id2label

{0: 'NEGATIVE', 1: 'POSITIVE'}

**Intentalo**:  Elija dos (o más) textos propios y póngalos a través del pipeline de análisis de sentimientos. A continuación, replique usted mismo los pasos que ha visto aquí y compruebe que obtiene los mismos resultados.


In [ ]:
import torch
from transformers import AutoTokenizer, AutoModelForSequenceClassification

checkpoint = "distilbert-base-uncased-finetuned-sst-2-english"
tokenizer = AutoTokenizer.from_pretrained(checkpoint)
model = AutoModelForSequenceClassification.from_pretrained(checkpoint)
sequences = ["I've been waiting for a Machine Learning course my whole life.", "So have I!"]

tokens = tokenizer(sequences, padding=True, truncation=True, return_tensors="pt")
output = model(**tokens)

/usr/local/lib/python3.10/dist-packages/transformers/tokenization_utils_base.py:1601: FutureWarning: `clean_up_tokenization_spaces` was not set. It will be set to `True` by default. This behavior will be depracted in transformers v4.45, and will be then set to `False` by default. For more details check this issue: https://github.com/huggingface/transformers/issues/31884
  warnings.warn(


In [ ]:
tokens

{'input_ids': tensor([[ 101, 1045, 1005, 2310, 2042, 3403, 2005, 1037, 3698, 4083, 2607, 2026,
         2878, 2166, 1012,  102],
        [ 101, 2061, 2031, 1045,  999,  102,    0,    0,    0,    0,    0,    0,
            0,    0,    0,    0]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
        [1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]])}

In [ ]:
output.logits

tensor([[ 2.8205, -2.4491],
        [-3.6183,  3.9137]], grad_fn=<AddmmBackward0>)

### Modelos

En esta sección, analizaremos más de cerca la creación y el uso de un modelo. Usaremos la clase AutoModel, que es útil cuando desea crear una instancia de cualquier modelo desde un punto de control.

La clase AutoModel y todos sus parientes son en realidad simples contenedores de la amplia variedad de modelos disponibles en la biblioteca. Es un contenedor inteligente, ya que puede adivinar automáticamente la arquitectura de modelo adecuada para su punto de control y, a continuación, crea una instancia de un modelo con esta arquitectura.

### Creando un transformer

In [ ]:
from transformers import BertConfig, BertModel

# Building the config
config = BertConfig()

# Building the model from the config
model = BertModel(config)

La configuración contiene muchos atributos que son usados para construir el modelo

In [ ]:
print(config)

BertConfig {
  "attention_probs_dropout_prob": 0.1,
  "classifier_dropout": null,
  "hidden_act": "gelu",
  "hidden_dropout_prob": 0.1,
  "hidden_size": 768,
  "initializer_range": 0.02,
  "intermediate_size": 3072,
  "layer_norm_eps": 1e-12,
  "max_position_embeddings": 512,
  "model_type": "bert",
  "num_attention_heads": 12,
  "num_hidden_layers": 12,
  "pad_token_id": 0,
  "position_embedding_type": "absolute",
  "transformers_version": "4.44.2",
  "type_vocab_size": 2,
  "use_cache": true,
  "vocab_size": 30522
}



Cargar un modelo de Transformer que ya está entrenado es simple: podemos hacer esto usando el método from_pretrained():

In [ ]:
from transformers import BertModel

model = BertModel.from_pretrained("bert-base-cased")

model.safetensors:   0%|          | 0.00/436M [00:00<?, ?B/s]

En el ejemplo de código anterior, no usamos BertConfig y, en su lugar, cargamos un modelo previamente entrenado a través del identificador bert-base-caseed. Este es un modelo de punto de control que fue entrenado por los propios autores de BERT.

In [ ]:
dir(model)

['T_destination',
 '__annotations__',
 '__call__',
 '__class__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattr__',
 '__getattribute__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__module__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__setstate__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 '__weakref__',
 '_apply',
 '_assisted_decoding',
 '_auto_class',
 '_autoset_attn_implementation',
 '_backward_compatibility_gradient_checkpointing',
 '_backward_hooks',
 '_backward_pre_hooks',
 '_beam_search',
 '_buffers',
 '_call_impl',
 '_check_and_enable_flash_attn_2',
 '_check_and_enable_sdpa',
 '_compiled_call_impl',
 '_constrained_beam_search',
 '_contrastive_search',
 '_convert_head_mask_to_5d',
 '_copy_lm_head_original_to_resized',
 '_create_repo',
 '_dispatch_accelerate_model',
 '_dola_decoding',
 '_expand_inputs_for_generation',

In [ ]:
print(model.config)

BertConfig {
  "_name_or_path": "bert-base-cased",
  "architectures": [
    "BertForMaskedLM"
  ],
  "attention_probs_dropout_prob": 0.1,
  "classifier_dropout": null,
  "gradient_checkpointing": false,
  "hidden_act": "gelu",
  "hidden_dropout_prob": 0.1,
  "hidden_size": 768,
  "initializer_range": 0.02,
  "intermediate_size": 3072,
  "layer_norm_eps": 1e-12,
  "max_position_embeddings": 512,
  "model_type": "bert",
  "num_attention_heads": 12,
  "num_hidden_layers": 12,
  "pad_token_id": 0,
  "position_embedding_type": "absolute",
  "transformers_version": "4.44.2",
  "type_vocab_size": 2,
  "use_cache": true,
  "vocab_size": 28996
}



Este modelo ahora se inicializa con todos los pesos del punto de control. Se puede usar directamente para inferir las tareas en las que se entrenó, y también se puede ajustar en una nueva tarea. Al entrenar con pesos preentrenados en lugar desde cero, podemos lograr rápidamente buenos resultados.

**Usando el modelo**

In [ ]:
sequences = ["Hello!", "Cool.", "Nice!"]

In [ ]:
encoded_sequences = [
    [101, 7592, 999, 102],
    [101, 4658, 1012, 102],
    [101, 3835, 999, 102],
]

Esta es una lista de listas: la secuencia codificada

In [ ]:
import torch

model_inputs = torch.tensor(encoded_sequences)

In [ ]:
output = model(model_inputs)

In [ ]:
output

BaseModelOutputWithPoolingAndCrossAttentions(last_hidden_state=tensor([[[ 4.4496e-01,  4.8276e-01,  2.7797e-01,  ..., -5.4032e-02,
           3.9393e-01, -9.4770e-02],
         [ 2.4943e-01, -4.4093e-01,  8.1772e-01,  ..., -3.1917e-01,
           2.2992e-01, -4.1172e-02],
         [ 1.3668e-01,  2.2518e-01,  1.4502e-01,  ..., -4.6915e-02,
           2.8224e-01,  7.5566e-02],
         [ 1.1789e+00,  1.6738e-01, -1.8187e-01,  ...,  2.4671e-01,
           1.0441e+00, -6.1972e-03]],

        [[ 3.6436e-01,  3.2464e-02,  2.0258e-01,  ...,  6.0110e-02,
           3.2451e-01, -2.0996e-02],
         [ 7.1866e-01, -4.8725e-01,  5.1740e-01,  ..., -4.4012e-01,
           1.4553e-01, -3.7545e-02],
         [ 3.3223e-01, -2.3271e-01,  9.4876e-02,  ..., -2.5268e-01,
           3.2172e-01,  8.1085e-04],
         [ 1.2523e+00,  3.5754e-01, -5.1320e-02,  ..., -3.7840e-01,
           1.0526e+00, -5.6255e-01]],

        [[ 2.4042e-01,  1.4718e-01,  1.2110e-01,  ...,  7.6062e-02,
           3.3564e-01,  2

### Tokenizador

In [ ]:
from transformers import AutoTokenizer

tokenizer = AutoTokenizer.from_pretrained("bert-base-cased")
tokenizer("Using a Transformer network is simple")

/usr/local/lib/python3.10/dist-packages/transformers/tokenization_utils_base.py:1601: FutureWarning: `clean_up_tokenization_spaces` was not set. It will be set to `True` by default. This behavior will be depracted in transformers v4.45, and will be then set to `False` by default. For more details check this issue: https://github.com/huggingface/transformers/issues/31884
  warnings.warn(


{'input_ids': [101, 7993, 170, 13809, 23763, 2443, 1110, 3014, 102], 'token_type_ids': [0, 0, 0, 0, 0, 0, 0, 0, 0], 'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1]}

#### Encoding

 La codificación se realiza en un proceso de dos pasos: la tokenización, seguida de la conversión a ID de entrada.

 Como hemos visto, el primer paso es dividir el texto en palabras (o partes de palabras, símbolos de puntuación, etc.), normalmente llamadas tokens. Hay varias reglas que pueden gobernar ese proceso, por lo que necesitamos crear una instancia del tokenizador con el nombre del modelo, para asegurarnos de que usamos las mismas reglas que se usaron cuando se entrenó previamente el modelo.

El segundo paso es convertir esos tokens en números, para que podamos construir un tensor a partir de ellos y alimentarlos al modelo. Para ello, el tokenizador cuenta con un vocabulario, que es la parte que descargamos cuando lo instanciamos con el método from_pretrained(). De nuevo, tenemos que usar el mismo vocabulario que se utilizó cuando se entrenó previamente el modelo.

In [ ]:
from transformers import AutoTokenizer

tokenizer = AutoTokenizer.from_pretrained("bert-base-cased")

sequence = "Using a Transformer network is simple"
tokens = tokenizer.tokenize(sequence)

print(tokens)

['Using', 'a', 'Trans', '##former', 'network', 'is', 'simple']


Obtener los ids

In [ ]:
ids = tokenizer.convert_tokens_to_ids(tokens)

print(ids)

[7993, 170, 13809, 23763, 2443, 1110, 3014]


**Decodificado**

In [ ]:
decoded_string = tokenizer.decode([7993, 170, 11303, 1200, 2443, 1110, 3014])
print(decoded_string)

Using a transformer network is simple


### Ejemplo

In [ ]:
!pip install datasets

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 527.3/527.3 kB 8.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 116.3/116.3 kB 9.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 39.9/39.9 MB 14.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 134.8/134.8 kB 11.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 194.1/194.1 kB 15.4 MB/s eta 0:00:00
  Attempting uninstall: pyarrow
    Found existing installation: pyarrow 14.0.2
    Uninstalling pyarrow-14.0.2:
      Successfully uninstalled pyarrow-14.0.2
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
cudf-cu12 24.4.1 requires pyarrow<15.0.0a0,>=14.0.1, but you have pyarrow 17.0.0 which is incompatible.
ibis-framework 8.0.0 requires pyarrow<16,>=2, but you have pyarrow 17.0.0 which is incompatible.


In [ ]:
from datasets import load_dataset

In [ ]:
dataset = load_dataset('rotten_tomatoes')

/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:89: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


Generating train split:   0%|          | 0/8530 [00:00<?, ? examples/s]

Generating validation split:   0%|          | 0/1066 [00:00<?, ? examples/s]

Generating test split:   0%|          | 0/1066 [00:00<?, ? examples/s]

In [ ]:
dataset['train'][0]

{'text': 'the rock is destined to be the 21st century\'s new " conan " and that he\'s going to make a splash even greater than arnold schwarzenegger , jean-claud van damme or steven segal .',
 'label': 1}

In [ ]:
from transformers import AutoTokenizer, TFAutoModelForSequenceClassification, DataCollatorWithPadding

In [ ]:
Tokenizer = AutoTokenizer.from_pretrained('bert-base-cased')

tokenizer_config.json:   0%|          | 0.00/49.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/570 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/213k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/436k [00:00<?, ?B/s]

/usr/local/lib/python3.10/dist-packages/transformers/tokenization_utils_base.py:1601: FutureWarning: `clean_up_tokenization_spaces` was not set. It will be set to `True` by default. This behavior will be depracted in transformers v4.45, and will be then set to `False` by default. For more details check this issue: https://github.com/huggingface/transformers/issues/31884
  warnings.warn(


In [ ]:
Tokenizer.tokenize(dataset['train'][0]['text'])

['the',
 'rock',
 'is',
 'destined',
 'to',
 'be',
 'the',
 '21st',
 'century',
 "'",
 's',
 'new',
 '"',
 'con',
 '##an',
 '"',
 'and',
 'that',
 'he',
 "'",
 's',
 'going',
 'to',
 'make',
 'a',
 'splash',
 'even',
 'greater',
 'than',
 'a',
 '##rno',
 '##ld',
 's',
 '##ch',
 '##war',
 '##zen',
 '##eg',
 '##ger',
 ',',
 'j',
 '##ean',
 '-',
 'c',
 '##lau',
 '##d',
 'van',
 'dam',
 '##me',
 'or',
 's',
 '##te',
 '##ven',
 'se',
 '##gal',
 '.']

In [ ]:
def tokenize_function(examples):
    return Tokenizer(examples["text"], padding="max_length", truncation=True)

In [ ]:
tokenized_datasets = dataset.map(tokenize_function, batched=True)

Map:   0%|          | 0/8530 [00:00<?, ? examples/s]

Map:   0%|          | 0/1066 [00:00<?, ? examples/s]

Map:   0%|          | 0/1066 [00:00<?, ? examples/s]

In [ ]:
tokenized_datasets

DatasetDict({
    train: Dataset({
        features: ['text', 'label', 'input_ids', 'token_type_ids', 'attention_mask'],
        num_rows: 8530
    })
    validation: Dataset({
        features: ['text', 'label', 'input_ids', 'token_type_ids', 'attention_mask'],
        num_rows: 1066
    })
    test: Dataset({
        features: ['text', 'label', 'input_ids', 'token_type_ids', 'attention_mask'],
        num_rows: 1066
    })
})

In [ ]:
data_collator = DataCollatorWithPadding(tokenizer=Tokenizer, return_tensors="tf")

In [ ]:
data_collator

DataCollatorWithPadding(tokenizer=BertTokenizerFast(name_or_path='bert-base-cased', vocab_size=28996, model_max_length=512, is_fast=True, padding_side='right', truncation_side='right', special_tokens={'unk_token': '[UNK]', 'sep_token': '[SEP]', 'pad_token': '[PAD]', 'cls_token': '[CLS]', 'mask_token': '[MASK]'}, clean_up_tokenization_spaces=True),  added_tokens_decoder={
	0: AddedToken("[PAD]", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
	100: AddedToken("[UNK]", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
	101: AddedToken("[CLS]", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
	102: AddedToken("[SEP]", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
	103: AddedToken("[MASK]", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
}, padding=True, max_length=None, pad_to_multiple_of=None, return_tensors='tf')

In [ ]:
tf_train_dataset = tokenized_datasets["train"].to_tf_dataset(
    columns=["attention_mask", "input_ids", "label"],
    label_cols=["label"],
    shuffle=True,
    collate_fn=data_collator,
    batch_size=8,
)

/usr/local/lib/python3.10/dist-packages/datasets/arrow_dataset.py:410: FutureWarning: The output of `to_tf_dataset` will change when a passing single element list for `labels` or `columns` in the next datasets version. To return a tuple structure rather than dict, pass a single string.
Old behaviour: columns=['a'], labels=['labels'] -> (tf.Tensor, tf.Tensor)  
             : columns='a', labels='labels' -> (tf.Tensor, tf.Tensor)  
New behaviour: columns=['a'],labels=['labels'] -> ({'a': tf.Tensor}, {'labels': tf.Tensor})  
             : columns='a', labels='labels' -> (tf.Tensor, tf.Tensor) 
  warnings.warn(


In [ ]:
!pip install huggingface-hub==0.23.2

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 401.7/401.7 kB 21.4 MB/s eta 0:00:00
  Attempting uninstall: huggingface-hub
    Found existing installation: huggingface-hub 0.24.6
    Uninstalling huggingface-hub-0.24.6:
      Successfully uninstalled huggingface-hub-0.24.6


In [ ]:
!pip install langchain-community
#0.2.16

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 50.4/50.4 kB 4.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.3/2.3 MB 61.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.0/1.0 MB 43.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 396.6/396.6 kB 27.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 289.8/289.8 kB 23.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 76.4/76.4 kB 7.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 77.9/77.9 kB 7.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 49.3/49.3 kB 4.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 141.9/141.9 kB 12.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 58.3/58.3 kB 5.5 MB/s eta 0:00:00
  Attempting uninstall: tenacity
    Found existing installation: tenacity 9.0.0
    Uninstalling tenacity-9.0.0:
      Successfully uninstalled tenacity-9.0.0


In [ ]:
pip install -U langchain-huggingface

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 249.1/249.1 kB 16.1 MB/s eta 0:00:00


In [ ]:
# LLM: HuggingFace
#from langchain_community.embeddings import HuggingFaceInferenceAPIEmbeddings
from langchain_huggingface import HuggingFaceEndpoint

In [ ]:
from google.colab import userdata
temperature = 0.5
top_p = 0.95
#model_name =   "EleutherAI/gpt-neo-1.3B"/ /"distilgpt2"
model_name =  "mistralai/Mistral-7B-Instruct-v0.2"
hugging_key = userdata.get('HF_TOKEN')

In [ ]:
llm = HuggingFaceEndpoint(
            repo_id=model_name,
            huggingfacehub_api_token=hugging_key,
            temperature = temperature,
            max_new_tokens = 50,
            top_p = top_p,
            timeout=400
        )

The token has not been saved to the git credentials helper. Pass `add_to_git_credential=True` in this function directly or `--add-to-git-credential` if using via `huggingface-cli` if you want to set the git credential as well.
Token is valid (permission: fineGrained).
Your token has been saved to /root/.cache/huggingface/token
Login successful


In [ ]:
# Configura el prompt de prueba
prompt = "Describe los beneficios del aprendizaje automático en la industria culinaria."

In [ ]:
# Genera la salida del modelo
response = llm(prompt)

<ipython-input-9-ddba49d97e76>:2: LangChainDeprecationWarning: The method `BaseLLM.__call__` was deprecated in langchain-core 0.1.7 and will be removed in 1.0. Use invoke instead.
  response = llm(prompt)


InferenceTimeoutError: Model not loaded on the server: https://api-inference.huggingface.co/models/mistralai/Mistral-7B-Instruct-v0.2. Please retry with a higher timeout (current: 400).

In [ ]:
response

'\n\nLa industria de la cocina es una de las industrias más importantes de la economía argentina y se ha convertido en uno de los sectores más productivos del país.\n\nEl sector de la cocina tiene una economía de $1,2 billones de dólares y se estima que el sector de la cocina de la Argentina se ha convertido en uno de los sectores'

In [ ]:
# Muestra la respuesta generada
print(response['generated_text'])

TypeError: string indices must be integers

In [ ]:
from transformers import pipeline

# Carga del modelo localmente
generator = pipeline('text-generation', model='EleutherAI/gpt-neo-1.3B')

# Generar texto
response = generator("Describe los beneficios del aprendizaje automático en la industria de la salud.",
                     max_length=1024,
                     do_sample=True,
                     temperature=0.7)
print(response[0]['generated_text'])


Truncation was not explicitly activated but `max_length` is provided a specific value, please use `truncation=True` to explicitly truncate examples to max length. Defaulting to 'longest_first' truncation strategy. If you encode pairs of sequences (GLUE-style) with the tokenizer you can select this strategy more precisely by providing a specific strategy to `truncation`.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


Describe los beneficios del aprendizaje automático en la industria de la salud.

En esta oportunidad, el Comité de Aprendizaje de la Sociedad Española de la Salud (CASE) dirige una conferencia sobre el aprendizaje automático en la industria de la salud.

CASE, en su informe sobre el aprendizaje automático en la industria de la salud, señala que existen muchas otras cuestiones más importantes que deben tenerse en cuenta en el aprendizaje automático.

En la conferencia, el Comité de Aprendizaje de la Sociedad Española de la Salud (CASE) señala que el aprendizaje automático puede ser una herramienta importante para la competitividad e integración del sistema sanitario.

En su informe sobre el aprendizaje automático en la industria de la salud, el Comité de Aprendizaje de la Sociedad Española de la Salud (CASE) señala que los beneficios del aprendizaje automático son de interés para la industria de la salud.

CASE, en su informe sobre el aprendizaje automático en la industria de la salud, 

In [ ]:
from transformers import AutoTokenizer, AutoModelForCausalLM

checkpoint = "meta-llama/Llama-2-7b-chat-hf"
tokenizer = AutoTokenizer.from_pretrained(checkpoint)
model = AutoModelForCausalLM.from_pretrained(checkpoint)

OSError: We couldn't connect to 'https://huggingface.co' to load this file, couldn't find it in the cached files and it looks like meta-llama/Llama-2-7b-chat-hf is not the path to a directory containing a file named config.json.
Checkout your internet connection or see how to run the library in offline mode at 'https://huggingface.co/docs/transformers/installation#offline-mode'.

In [ ]:
# Preparar el prompt
prompt = "Explica de manera simple que es la inteligencia artificial"

#Tokenizar el input
inputs = tokenizer(prompt, return_tensors="pt")

# Generar el output
output = model.generate(inputs.inputs_ids, max_new_tokens=200)

#Decodificar e imprimir la respuesta
response = tokenizer.decode(output[0], skip_special_tokens=True)
print(response)